lib/librte_hash/rte_cuckoo_hash.c

```diff
@@ -1339,6 +1339,215 @@ rte_hash_lookup_with_hash(const struct rte_hash *h,
+		 * This can result in the reader loading old signature
+		 * (which matches), while the key_idx is updated to a
+		 * value that belongs to a new key. However, the full
+		 * key comparison will ensure that the lookup fails.
+		 */
+		if (bkt->sig_current[i] == sig) {
+			key_idx = __atomic_load_n(&bkt->key_idx[i],
+					  __ATOMIC_ACQUIRE);
+			if (key_idx != EMPTY_SLOT) {
+				k = (struct rte_hash_key *) ((char *)keys +
+						key_idx * h->key_entry_size);
+				yield_func(k, arg);
+
+				if (rte_hash_cmp_eq(key, k->key, h) == 0) {
+					if (data != NULL) {
+						*data = __atomic_load_n(
+							&k->pdata,
+							__ATOMIC_ACQUIRE);
+					}
+					/*
+					 * Return index where key is stored,
+					 * subtracting the first dummy index
+					 */
+					return key_idx - 1;
+				}
+			}
+		}
+	}
+	return -1;
+}
+
+static inline int32_t
+__rte_hash_lookup_and_yield_with_hash_l(const struct rte_hash *h,
+	const void *key, hash_sig_t sig, void **data,
+	rte_hash_yield_func yield_func, void *arg)
+{
+	uint32_t prim_bucket_idx, sec_bucket_idx;
+	struct rte_hash_bucket *bkt, *cur_bkt;
+	int ret;
+	uint16_t short_sig;
+
+	short_sig = get_short_sig(sig);
+	prim_bucket_idx = get_prim_bucket_index(h, sig);
+	sec_bucket_idx = get_alt_bucket_index(h, prim_bucket_idx, short_sig);
+
+	bkt = &h->buckets[prim_bucket_idx];
+	yield_func(bkt, arg);
+
+	__hash_rw_reader_lock(h);
+
+	/* Check if key is in primary location */
+	ret = search_one_bucket_and_yield_l(h, key, short_sig, data, bkt,
+		yield_func, arg);
+	if (ret != -1) {
+		__hash_rw_reader_unlock(h);
+		return ret;
+	}
+	/* Calculate secondary hash */
+	bkt = &h->buckets[sec_bucket_idx];
+
+	/* Check if key is in secondary location */
+	FOR_EACH_BUCKET(cur_bkt, bkt) {
+		yield_func(cur_bkt, arg);
+		ret = search_one_bucket_and_yield_l(h, key, short_sig,
+			data, cur_bkt, yield_func, arg);
+		if (ret != -1) {
+			__hash_rw_reader_unlock(h);
+			return ret;
+		}
+	}
+
+	__hash_rw_reader_unlock(h);
+
+	return -ENOENT;
+}
+
+static inline int32_t
+__rte_hash_lookup_and_yield_with_hash_lf(const struct rte_hash *h,
+	const void *key, hash_sig_t sig, void **data,
+	rte_hash_yield_func yield_func, void *arg)
+{
+	uint32_t prim_bucket_idx, sec_bucket_idx;
+	struct rte_hash_bucket *bkt, *cur_bkt;
+	uint32_t cnt_b, cnt_a;
+	int ret;
+	uint16_t short_sig;
+
+	short_sig = get_short_sig(sig);
+	prim_bucket_idx = get_prim_bucket_index(h, sig);
+	sec_bucket_idx = get_alt_bucket_index(h, prim_bucket_idx, short_sig);
+
+	do {
+		/* Load the table change counter before the lookup
+		 * starts. Acquire semantics will make sure that
+		 * loads in search_one_bucket are not hoisted.
+		 */
+		cnt_b = __atomic_load_n(h->tbl_chng_cnt,
+				__ATOMIC_ACQUIRE);
+
+		/* Check if key is in primary location */
+		bkt = &h->buckets[prim_bucket_idx];
+		yield_func(bkt, arg);
+		ret = search_one_bucket_and_yield_lf(h, key, short_sig, data,
+			bkt, yield_func, arg);
+		if (ret != -1) {
+			__hash_rw_reader_unlock(h);
```

Comment:
I know that you copied the line above from DPDK, but it seems to be a bug in DPDK here: this function is lock-free and there's no call to `__hash_rw_reader_lock()`. Could you verify if the latest version of DPDK has fixed it? If you confirm this is a bug and it hasn't been fixed in the latest version of DPDK, it's worth reporting.

This comment also applies to the call `__hash_rw_reader_unlock()` below.

Reply:
After taking a look at [the latest version of DPDK](https://git.dpdk.org/dpdk/tree/lib/librte_hash/rte_cuckoo_hash.c#n1264), it doesn't change this call.

Reply:
``__hash_rw_reader_lock()`` is called by ``__rte_hash_lookup_with_hash_l()``, ``__rte_hash_lookup_bulk_l()`` and ``rte_hash_iterate()``, so when ``rte_hash_iterate()`` is called by one thread (e.g., ``t1``) which obtained a reader lock and is reading the item with key ``k1``,  another thread calling ``__rte_hash_lookup_with_hash_lf()`` releases a reader lock due to a failure lookup. At the same time, one thread calling either deletion or addition can successfully get the writer lock and write an item with key ``k1`` into the hash table while thread ``t1`` is accessing the item, then collision happens!

Reply:
Thank you for the investigation.